[RFC] docs: deprecation guideline (#986)
DivanteLtd/storefront-ui: packages/vue/docs/contributing/coding-guidelines.md

Try to make components as customizable as possible **without complicating** them. Think about the parts that are usually customized and allow simple way to change their look.
:::

## Deprecation guidelines

> StorefrontUI cares about providing new features and keeping up-to-date with new best practices. Sometimes, it can be necessary to make some changes in APIs or prepare new features to stay current. To make this process easier for everyone, we deprecate APIs and features for a period before actual removal, by adding proper info about deprecation in the next major version.

Reviewer: ```suggestion
StorefrontUI cares about providing new features and keeping up-to-date with new best practices. Sometimes it can become necessary to make changes to APIs or to prepare new features in order to stay current. To make this process easier for everyone, we deprecate APIs and features for a period of time before the actual removal by adding proper information about deprecations in the next major version.
```

> **In result:**
> - we don't remove implemented features before next major version
> - we add info about deprecation, so user can be prepared for substitute and have time to update package to the latest version
> - before release, we can easily find every deprecated feature -> remove it and prepare documentation for it

Reviewer: ```suggestion
**In summary:**
- we don't remove implemented features before the next major version
- we add deprecation infos, so users can prepare for replacements and have time to update their dependencies to the latest version
- on our side: before a release, we can spot deprecated features more easily, remove them and prepare migration documentation
```

> **To standarize these process, we can stick with few rules:**

Reviewer: ```suggestion
**To standardize this process, we stick with a few rules:**
```

- add this note next to the deprecated feature/APIs -> `@deprecated will be removed in 1.0.0 use ___ prop instead`
- add this note next to deprecated code that should be removed `// TODO remove in 1.0.0 / BEGIN` -> `// END`
- import helper from utitilies `import { deprecationWarning } from "@/utilities/helpers/deprecation-warning.js";`

> - add info about deprecation by using method `deprecationWarning(this.$options.name, "yourMessage")`

Reviewer: ```suggestion
- add info about deprecation by using the helper method `deprecationWarning(this.$options.name, "your deprecation message")`
```

:::warning deprecationWarning(componentName, message)

> componentName is an argument that you have to pass to put name of component with depreacted features - `this.$options.name`

Reviewer: ```suggestion
`componentName` is a string argument that you have to pass for the component's name to appear in front of the deprecation warning – in general, `this.$options.name` returns the name of the current Vue component.
```

:::

> In result, console will show warnining:

Reviewer: ```suggestion
As a result the console will show a warning like the following:
```

 `[StorefrontUI][SfComponent] ____ prop has been deprecated and will be removed in 1.0.0. Use ____ instead.`

:::tip DEPRECATION INFO
Here you can see an example how it looks like in SfAccordion:
::: 

```js
props: {
  /**
   * Opens the first accordion item if set to "true"
   * @deprecated will be removed in 1.0.0 use open prop instead
   */
  firstOpen: {
    type: Boolean,
    default: false
  },
}
```

```js
// TODO remove in 1.0.0 / BEGIN
if (this.firstOpen) {
  this.$children[0].isOpen = this.firstOpen;
  console.warn(

>     "[StorefrontUI][SfAccordion] firstOpen prop has been deprecated and will be removed in 1.0.0. Use open instead."

Reviewer: ```suggestion
    "[StorefrontUI][SfAccordion] Prop 'firstOpen' has been deprecated and will be removed in v1.0.0. Use 'open' instead."
```
